File: backend/routers/hazards.py
from fastapi import APIRouter, Query, HTTPException
from pydantic import BaseModel
from typing import Optional
import database

router = APIRouter(tags=["hazards"])


class HazardIn(BaseModel):
    latitude: float
    longitude: float
    confidence: float
    hazard_type: str
    severity: str = "medium"
    session_id: Optional[str] = None


@router.get("/hazards")
def get_hazards(
    hazard_type: Optional[str] = Query(None),
    severity: Optional[str] = Query(None),
    since_hours: Optional[int] = Query(None),
):
    return {"hazards": database.list_hazards(hazard_type, severity, since_hours)}


@router.get("/hazards/{hazard_id}")
def get_hazard(hazard_id: int):
    h = database.get_hazard(hazard_id)
    if not h:
        raise HTTPException(status_code=404, detail="Hazard not found")
    return h


@router.post("/hazards")
def upload_hazard(h: HazardIn):
    nearby_id = database.find_nearby_hazard(h.latitude, h.longitude, threshold_m=50, hazard_type=h.hazard_type)
    if nearby_id:
        database.update_hazard(nearby_id, h.latitude, h.longitude, h.confidence)
        return {"status": "merged", "hazard_id": nearby_id}
    new_id = database.insert_hazard(h.latitude, h.longitude, h.hazard_type, h.confidence, h.severity)
    return {"status": "created", "hazard_id": new_id}

File: frontend/src/api/hazards.ts
import { api } from './client';
import type { Hazard, Severity, HazardType } from './types';

export function fetchHazards(params?: {
  hazard_type?: HazardType;
  severity?: Severity;
  since_hours?: number;
}) {
  const qs = new URLSearchParams();
  if (params?.hazard_type) qs.set('hazard_type', params.hazard_type);
  if (params?.severity) qs.set('severity', params.severity);
  if (params?.since_hours) qs.set('since_hours', String(params.since_hours));
  return api.get<{ hazards: Hazard[] }>(`/api/hazards?${qs}`);
}

export function uploadHazard(payload: {
  latitude: number;
  longitude: number;
  confidence: number;
  hazard_type: string;
  severity: string;
  session_id?: string;
}) {
  return api.post<{ status: string; hazard_id: number }>('/api/hazards', payload);
}
